src/x_net.c

```diff
@@ -278,48 +284,86 @@ static void *netsend_new(t_symbol *s, int argc, t_atom *argv)
+    int ret = 0, readbytes = 0, i;
     struct sockaddr_storage fromaddr = {0};
     socklen_t fromaddrlen = sizeof(struct sockaddr_storage);
     if (!x->x_msgout)
     {
         bug("netsend_readbin");
         return;
     }
-    if (x->x_protocol == SOCK_DGRAM)
-        ret = (int)recvfrom(fd, inbuf, MAXPDSTRING, 0,
-            (struct sockaddr *)&fromaddr, &fromaddrlen);
-    else
-        ret = (int)recv(fd, inbuf, MAXPDSTRING, 0);
-    if (ret <= 0)
-    {
-        if (ret < 0)
-            sys_sockerror("recv (bin)");
-        sys_rmpollfn(fd);
-        sys_closesocket(fd);
-        if (x->x_obj.ob_pd == netreceive_class)
-            netreceive_notify((t_netreceive *)x, fd);
-        return;
-    }
-    if (x->x_protocol == SOCK_DGRAM)
-    {
-        if (x->x_fromout)
-            outlet_sockaddr(x->x_fromout, (const struct sockaddr *)&fromaddr);
-        t_atom *ap = (t_atom *)alloca(ret * sizeof(t_atom));
-        for (i = 0; i < ret; i++)
-            SETFLOAT(ap+i, inbuf[i]);
-        outlet_list(x->x_msgout, 0, ret, ap);
-    }
-    else
+    while (1)
     {
-        if (x->x_fromout &&
-            !getpeername(fd, (struct sockaddr *)&fromaddr, &fromaddrlen))
+        if (x->x_protocol == SOCK_DGRAM)
+            ret = (int)recvfrom(fd, inbuf, INBUFSIZE, 0,
+                (struct sockaddr *)&fromaddr, &fromaddrlen);
+        else
+            ret = (int)recv(fd, inbuf, INBUFSIZE, 0);
+        if (ret <= 0)
+        {
+            if (ret < 0)
+            {
+                /* only close the socket if there really was an error.
+                (sys_sockerrno() ignores some error codes) */
+                if (!sys_sockerrno())
+                    return;
+                sys_sockerror("recv (bin)");
+            }
+            if (x->x_obj.ob_pd == netreceive_class)
+            {
+                sys_rmpollfn(fd);
+                sys_closesocket(fd);
+                netreceive_notify((t_netreceive *)x, fd);
+            }
+            else /* properly shutdown netsend */
+                netsend_disconnect(x);
+            return;
+        }
+        if (x->x_protocol == SOCK_DGRAM)
+        {
+            t_atom *ap;
+        #ifdef _WIN32
+            unsigned long n;
+        #else
+            int n;
+        #endif
+            if (x->x_fromout)
                 outlet_sockaddr(x->x_fromout, (const struct sockaddr *)&fromaddr);
-        for (i = 0; i < ret; i++)
-            outlet_float(x->x_msgout, inbuf[i]);
+            ap = (t_atom *)alloca(ret * sizeof(t_atom));
+            for (i = 0; i < ret; i++)
+                SETFLOAT(ap+i, inbuf[i]);
+            outlet_list(x->x_msgout, 0, ret, ap);
+            readbytes += ret;
+                /* throttle */
+            if (readbytes >= INBUFSIZE)
+                return;
+                /* check for pending UDP packets */
+        #ifdef _WIN32
+            if (ioctlsocket(fd, FIONREAD, &n) < 0)
+        #else
+            if (ioctl(fd, FIONREAD, &n) < 0)
```

Comment:
FYI: This doesn't actually work as FIONREAD doesn't work with ioctl(). You have to use fcntl() which is why I wrote the socket_set_nonblocking() wrapper.

Reply:
Ah wait, I'm wrong. The flag exists but is not used to set the socket in non-blocking mode for *nix.